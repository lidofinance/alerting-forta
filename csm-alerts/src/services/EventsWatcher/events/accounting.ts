import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { CSAccounting__factory } from '../../../generated/typechain'
import * as CSAccounting from '../../../generated/typechain/CSAccounting'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress, formatEther } from '../../../utils/string'

const ICSAccounting = CSAccounting__factory.createInterface()

export function getCSAccountingEvents(accounting: string): EventOfNotice[] {
    return [
        {
            address: accounting,
            abi: ICSAccounting.getEvent('ChargePenaltyRecipientSet').format('full'),
            alertId: 'CS-ACCOUNTING-CHARGE-PENALTY-RECIPIENT-SET',
            name: 'ðŸš¨ CSAccounting: Charge penalty recipient set',
            description: (args: CSAccounting.ChargePenaltyRecipientSetEvent.OutputObject) =>
                `Charge penalty recipient set to ${etherscanAddress(args.chargePenaltyRecipient)} (expecting the treasury)`,
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
        {
            address: accounting,
            abi: ICSAccounting.getEvent('BondCurveUpdated').format('full'),
            alertId: 'CS-ACCOUNTING-BOND-CURVE-UPDATED',
            name: 'ðŸš¨ CSAccounting: Bond curve updated',
            description: (args: CSAccounting.BondCurveUpdatedEvent.OutputObject) => {
                const bondCurveString = args.bondCurve
                    .map((value: bigint) => formatEther(value))
                    .join(', ')
                return `Bond curve #${args.curveId} updated. New values: [${bondCurveString}]`
            },
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
        {
            address: accounting,
            abi: ICSAccounting.getEvent('BondCurveAdded').format('full'),
            alertId: 'CS-ACCOUNTING-BOND-CURVE-ADDED',
            name: 'ðŸ”´ CSAccounting: Bond curve added',
            description: (args: CSAccounting.BondCurveAddedEvent.OutputObject) => {
                const bondCurveString = args.bondCurve
                    .map((value: bigint) => formatEther(value))
                    .join(', ')
                return `Bond curve added: [${bondCurveString}]`
            },
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
    ]
}
