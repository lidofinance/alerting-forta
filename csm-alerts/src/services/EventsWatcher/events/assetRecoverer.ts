import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { AssetRecoverer__factory } from '../../../generated/typechain'
import * as AssetRecoverer from '../../../generated/typechain/AssetRecoverer'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress, formatEther, formatShares } from '../../../utils/string'

const IAssetRecoverer = AssetRecoverer__factory.createInterface()

export function getAssetRecovererEvents(
    contracts: { name: string; address: string }[],
): EventOfNotice[] {
    return contracts.flatMap((contract) => {
        return [
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('ERC20Recovered').format('full'),
                alertId: 'ASSET-RECOVERER-ERC20-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: ERC20 recovered',
                description: (args: AssetRecoverer.ERC20RecoveredEvent.OutputObject) =>
                    `ERC20 recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Token: ${etherscanAddress(args.token)}\n` +
                    `Amount: ${args.amount}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('ERC721Recovered').format('full'),
                alertId: 'ASSET-RECOVERER-ERC721-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: ERC721 recovered',
                description: (args: AssetRecoverer.ERC721RecoveredEvent.OutputObject) =>
                    `ERC721 recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Token: ${etherscanAddress(args.token)}\n` +
                    `Token ID: ${args.tokenId}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('ERC1155Recovered').format('full'),
                alertId: 'ASSET-RECOVERER-ERC1155-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: ERC1155 recovered',
                description: (args: AssetRecoverer.ERC1155RecoveredEvent.OutputObject) =>
                    `ERC1155 recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Token: ${etherscanAddress(args.token)}\n` +
                    `Token ID: ${args.tokenId}\n` +
                    `Amount: ${args.amount}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('EtherRecovered').format('full'),
                alertId: 'ASSET-RECOVERER-ETHER-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: Ether recovered',
                description: (args: AssetRecoverer.EtherRecoveredEvent.OutputObject) =>
                    `Ether recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Amount: ${formatEther(args.amount)}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('StETHSharesRecovered').format('full'),
                alertId: 'ASSET-RECOVERER-STETH-SHARES-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: stETH Shares recovered',
                description: (args: AssetRecoverer.StETHSharesRecoveredEvent.OutputObject) =>
                    `StETH Shares recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Amount: ${formatShares(args.shares)}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
        ]
    })
}
