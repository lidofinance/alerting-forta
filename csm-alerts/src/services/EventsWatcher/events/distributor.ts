import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { CSFeeDistributor__factory } from '../../../generated/typechain'
import * as CSFeeDistributor from '../../../generated/typechain/CSFeeDistributor'
import { EventOfNotice } from '../../../shared/types'
import { ipfsLink } from '../../../utils/string'

const ICSFeeDistributor = CSFeeDistributor__factory.createInterface()

export function getCSFeeDistributorEvents(distributorAddress: string): EventOfNotice[] {
    return [
        {
            address: distributorAddress,
            abi: ICSFeeDistributor.getEvent('DistributionDataUpdated').format('full'),
            alertId: 'CSFEE-DISTRIBUTOR-DISTRIBUTION-DATA-UPDATED',
            name: 'ðŸ”µ CSFeeDistributor: Distribution data updated',
            description: (args: CSFeeDistributor.DistributionDataUpdatedEvent.OutputObject) =>
                `Total Claimable Shares: ${args.totalClaimableShares}\n` +
                `Tree Root: ${args.treeRoot}\n` +
                `Tree CID: ${ipfsLink(args.treeCid)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
        {
            address: distributorAddress,
            abi: ICSFeeDistributor.getEvent('DistributionLogUpdated').format('full'),
            alertId: 'CSFEE-DISTRIBUTOR-DISTRIBUTION-LOG-UPDATED',
            name: 'ðŸ”µ CSFeeDistributor: Distribution log updated',
            description: (args: CSFeeDistributor.DistributionLogUpdatedEvent.OutputObject) =>
                `Log CID: ${ipfsLink(args.logCid)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
    ]
}
