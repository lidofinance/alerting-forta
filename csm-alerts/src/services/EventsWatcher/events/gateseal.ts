import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { GateSeal__factory } from '../../../generated/typechain'
import * as GateSeal from '../../../generated/typechain/GateSeal'
import { EventOfNotice } from '../../../shared/types'
import { addressOnExplorer } from '../../../utils/string'
import { formatDelay } from '../../../utils/time'

const IGateSeal = GateSeal__factory.createInterface()

export function getGateSealEvents(
    address: string,
    knownContracts: { [key: string]: string },
): EventOfNotice[] {
    return [
        {
            address: address,
            abi: IGateSeal.getEvent('Sealed').format('full'),
            alertId: 'CS-GATE-SEAL-SEALED',
            name: 'ðŸš¨ CSM Gate Seal paused a contract',
            description: (args: GateSeal.SealedEvent.OutputObject) =>
                `${addressOnExplorer(args.sealable)} (${knownContracts[args.sealable] ?? 'unknown'})` +
                `was paused for ${formatDelay(args.sealed_for)} by ${addressOnExplorer(args.sealed_by)}`,
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
    ]
}
