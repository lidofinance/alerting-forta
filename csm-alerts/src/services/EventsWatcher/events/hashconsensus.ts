import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { HashConsensus__factory } from '../../../generated/typechain'
import * as HashConsensus from '../../../generated/typechain/HashConsensus'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress } from '../../../utils/string'

const IHashConsensus = HashConsensus__factory.createInterface()

export function getHashConsensusEvents(
    address: string,
    knownMembers: { [key: string]: string },
): EventOfNotice[] {
    return [
        {
            address,
            abi: IHashConsensus.getEvent('MemberAdded').format('full'),
            alertId: 'HASH-CONSENSUS-MEMBER-ADDED',
            name: 'ðŸ”´ CSM HashConsensus: Member added',
            description: (args: HashConsensus.MemberAddedEvent.OutputObject) =>
                `New member ${etherscanAddress(args.addr)} (${knownMembers[args.addr] ?? 'unknown'}) added\n` +
                `Total members: ${args.newTotalMembers}\n` +
                `New quorum: ${args.newQuorum}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('MemberRemoved').format('full'),
            alertId: 'HASH-CONSENSUS-MEMBER-REMOVED',
            name: 'ðŸ”´ CSM HashConsensus: Member removed',
            description: (args: HashConsensus.MemberRemovedEvent.OutputObject) =>
                `Member ${etherscanAddress(args.addr)} (${knownMembers[args.addr] ?? 'unknown'}) removed\n` +
                `Total members: ${args.newTotalMembers}\n` +
                `New quorum: ${args.newQuorum}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('QuorumSet').format('full'),
            alertId: 'HASH-CONSENSUS-QUORUM-SET',
            name: 'ðŸ”´ CSM HashConsensus: Quorum set',
            description: (args: HashConsensus.QuorumSetEvent.OutputObject) =>
                `Quorum set to ${args.newQuorum}.\n` +
                `Total members: ${args.totalMembers}\n` +
                `Previous quorum: ${args.prevQuorum}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('FastLaneConfigSet').format('full'),
            alertId: 'HASH-CONSENSUS-FASTLANE-CONFIG-SET',
            name: 'ðŸ”´ CSM HashConsensus: Fastlane config set',
            description: (args: HashConsensus.FastLaneConfigSetEvent.OutputObject) =>
                `Fastlane configuration set with length slots: ${args.fastLaneLengthSlots}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('FrameConfigSet').format('full'),
            alertId: 'HASH-CONSENSUS-FRAME-CONFIG-SET',
            name: 'ðŸ”´ CSM HashConsensus: Frame config set',
            description: (args: HashConsensus.FrameConfigSetEvent.OutputObject) =>
                `Frame configuration set:\n` +
                `New initial epoch: ${args.newInitialEpoch}\n` +
                `Epochs per frame: ${args.newEpochsPerFrame}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('ReportProcessorSet').format('full'),
            alertId: 'HASH-CONSENSUS-REPORT-PROCESSOR-SET',
            name: 'ðŸ”´ CSM HashConsensus: Report processor set',
            description: (args: HashConsensus.ReportProcessorSetEvent.OutputObject) =>
                `Previous processor: ${etherscanAddress(args.prevProcessor)}\n` +
                `Current processor: ${etherscanAddress(args.processor)}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('ConsensusLost').format('full'),
            alertId: 'HASH-CONSENSUS-LOST',
            name: 'ðŸ”´ CSM HashConsensus: Consensus lost',
            description: (args: HashConsensus.ConsensusLostEvent.OutputObject) =>
                `Consensus lost for slot ${args.refSlot}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('ConsensusReached').format('full'),
            alertId: 'HASH-CONSENSUS-REACHED',
            name: 'ðŸ”µ CSM HashConsensus: Consensus reached',
            description: (args: HashConsensus.ConsensusReachedEvent.OutputObject) =>
                `Consensus reached for slot ${args.refSlot}\n` +
                `Report hash: ${args.report}\n` +
                `Support: ${args.support}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
    ]
}
