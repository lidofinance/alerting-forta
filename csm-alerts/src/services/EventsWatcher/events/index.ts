export * from './assetRecoverer'
export * from './distributor'
export * from './accounting'
export * from './pausable'
export * from './proxies'
export * from './module'
export * from './oracle'
export * from './roles'
