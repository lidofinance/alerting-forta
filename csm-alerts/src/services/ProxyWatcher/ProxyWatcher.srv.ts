import { elapsedTime } from '../../utils/time'
import { either as E } from 'fp-ts'
import { Logger } from 'winston'
import { BlockDto, EventOfNotice, TransactionDto, handleEventsOfNotice } from '../../entity/events'
import { Finding } from '../../generated/proto/alert_pb'

export abstract class IProxyWatcherClient {
  public abstract getBlockByNumber(blockNumber: number): Promise<E.Either<Error, BlockDto>>
}

export class ProxyWatcherSrv {
  private readonly name = 'ProxyWatcherSrv'
  private readonly logger: Logger
  private readonly proxyWatcherClient: IProxyWatcherClient

  private readonly ossifiedProxyEvents: EventOfNotice[]
  private readonly pausableEvents: EventOfNotice[]
  private readonly assetRecovererEvents: EventOfNotice[]

  constructor(
    logger: Logger,
    ethProvider: IProxyWatcherClient,
    ossifiedProxyEvents: EventOfNotice[],
    pausableEvents: EventOfNotice[],
    assetRecovererEvents: EventOfNotice[],
  ) {
    this.logger = logger
    this.proxyWatcherClient = ethProvider

    this.ossifiedProxyEvents = ossifiedProxyEvents
    this.pausableEvents = pausableEvents
    this.assetRecovererEvents = assetRecovererEvents
  }

  public async initialize(currentBlock: number): Promise<Finding[] | null | Error> {
    const start = new Date().getTime()

    const currBlock = await this.proxyWatcherClient.getBlockByNumber(currentBlock)
    if (E.isLeft(currBlock)) {
      this.logger.error(elapsedTime(`Failed [${this.name}.initialize]`, start))
      return currBlock.left
    }
    this.logger.info(elapsedTime(`[${this.name}.initialize]`, start))
    return null
  }

  async handleBlock(blockDto: BlockDto): Promise<Finding[]> {
    const start = new Date().getTime()
    const findings: Finding[] = []

    this.logger.info(`${blockDto.number}`)
    this.logger.info(elapsedTime(ProxyWatcherSrv.name + '.' + this.handleBlock.name, start))

    return findings
  }

  public getName(): string {
    return this.name
  }

  public handleTransaction(txEvent: TransactionDto): Finding[] {
    const out: Finding[] = []

    const ossifiedProxyFindings = handleEventsOfNotice(txEvent, this.ossifiedProxyEvents)
    const pausableEventsFindings = handleEventsOfNotice(txEvent, this.pausableEvents)
    const assetRecovererFindings = handleEventsOfNotice(txEvent, this.assetRecovererEvents)

    out.push(...ossifiedProxyFindings, ...pausableEventsFindings, ...assetRecovererFindings)

    return out
  }
}
