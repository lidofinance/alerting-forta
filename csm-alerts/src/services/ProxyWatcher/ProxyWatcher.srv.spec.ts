import {
  CONTRACTS_WITH_ASSET_RECOVERER,
  CSM_PROXY_CONTRACTS,
  DeploymentAddress,
  DeploymentAddresses,
  PAUSABLE_CONTRACTS,
} from '../../utils/constants.holesky'
import { expect } from '@jest/globals'
import { TransactionDto } from '../../entity/events'
import {
  CSModule__factory,
  CSAccounting__factory,
  CSFeeDistributor__factory,
  CSFeeOracle__factory,
} from '../../generated/typechain'
import { getOssifiedProxyEvents } from '../../utils/events/ossified_proxy_events'
import { getPausableEvents } from '../../utils/events/pausable_events'
import { getAssetRecovererEvents } from '../../utils/events/asset_recoverer_events'
import { ProxyWatcherSrv, IProxyWatcherClient } from './ProxyWatcher.srv'
import * as Winston from 'winston'
import { ETHProvider } from '../../clients/eth_provider'
import { ethers } from 'forta-agent'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'
import promClient from 'prom-client'
import { Metrics } from '../../utils/metrics/metrics'

const TEST_TIMEOUT = 120_000 // ms

describe('ProxyWatcher event tests', () => {
  const chainId = 17000

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const address: DeploymentAddress = DeploymentAddresses

  const fortaEthersProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, chainId)
  const csModuleRunner = CSModule__factory.connect(address.CS_MODULE_ADDRESS, fortaEthersProvider)
  const csAccountingRunner = CSAccounting__factory.connect(address.CS_ACCOUNTING_ADDRESS, fortaEthersProvider)
  const csFeeDistributorRunner = CSFeeDistributor__factory.connect(
    address.CS_FEE_DISTRIBUTOR_ADDRESS,
    fortaEthersProvider,
  )
  const csFeeOracleRunner = CSFeeOracle__factory.connect(address.CS_FEE_ORACLE_ADDRESS, fortaEthersProvider)

  const registry = new promClient.Registry()
  const m = new Metrics(registry, 'test_')

  const proxyWatcherClient: IProxyWatcherClient = new ETHProvider(
    logger,
    m,
    fortaEthersProvider,
    csModuleRunner,
    csAccountingRunner,
    csFeeDistributorRunner,
    csFeeOracleRunner,
  )

  const proxyWatcherSrv = new ProxyWatcherSrv(
    logger,
    proxyWatcherClient,
    getOssifiedProxyEvents(CSM_PROXY_CONTRACTS),
    getPausableEvents(PAUSABLE_CONTRACTS),
    getAssetRecovererEvents(CONTRACTS_WITH_ASSET_RECOVERER),
  )

  test(
    'ðŸš¨ Proxy Watcher: Admin Changed',
    async () => {
      const txHash = '0x92410350f567757d8f73b2f4b3670454af3899d095103ea0e745c92714673277'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
      }

      const results = proxyWatcherSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(4)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸš¨ Proxy Watcher: Implementation Upgraded',
    async () => {
      const txHash = '0x262faac95560f7fc0c831580d17e48daa69b17831b798e0b00bc43168a310c52'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
      }

      const results = proxyWatcherSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(4)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸš¨ Proxy Watcher: Paused',
    async () => {
      const txHash = '0x56a49219b4e40d146c0dc11d795b6d43696947f0ceb63fad7e9b820eab2b3e14'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
      }

      const results = proxyWatcherSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(3)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸš¨ Proxy Watcher: Resumed',
    async () => {
      const txHash = '0xa44ac96956f254fe1b9d6ff0a60ad2b5b4a7eaff951eb98150c0f7bbe90dc5df'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
      }

      const results = proxyWatcherSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(3)
    },
    TEST_TIMEOUT,
  )
})
