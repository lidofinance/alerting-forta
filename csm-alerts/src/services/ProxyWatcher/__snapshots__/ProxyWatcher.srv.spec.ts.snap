// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProxyWatcher event tests ðŸŸ£ Admin Changed 1`] = `
[
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "PROXY-ADMIN-CHANGED",
      "ðŸŸ£ CSModule: Admin Changed",
      "The proxy admin for CSModule(0x4562c3e63c2e586cD1651B958C22F88135aCAd4f) has been changed from [0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164](https://etherscan.io/address/0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164) to [0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d](https://etherscan.io/address/0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164,0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "PROXY-ADMIN-CHANGED",
      "ðŸŸ£ CSAccounting: Admin Changed",
      "The proxy admin for CSAccounting(0xc093e53e8F4b55A223c18A2Da6fA00e60DD5EFE1) has been changed from [0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164](https://etherscan.io/address/0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164) to [0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d](https://etherscan.io/address/0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164,0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "PROXY-ADMIN-CHANGED",
      "ðŸŸ£ CSFeeOracle: Admin Changed",
      "The proxy admin for CSFeeOracle(0xaF57326C7d513085051b50912D51809ECC5d98Ee) has been changed from [0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164](https://etherscan.io/address/0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164) to [0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d](https://etherscan.io/address/0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0xc4DAB3a3ef68C6DFd8614a870D64D475bA44F164,0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`ProxyWatcher event tests ðŸŸ£ Implementation Upgraded 1`] = `
[
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "PROXY-UPGRADED",
      "ðŸŸ£ CSModule: Implementation Upgraded",
      "The proxy implementation has been upgraded to 0x4d70efa74ec0ac3a5f759cc0f714c94cbc5cc4da",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0x4D70efa74ec0Ac3a5f759cC0F714c94CBc5cc4da",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "PROXY-UPGRADED",
      "ðŸŸ£ CSAccounting: Implementation Upgraded",
      "The proxy implementation has been upgraded to 0x4d70efa74ec0ac3a5f759cc0f714c94cbc5cc4da",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0x4D70efa74ec0Ac3a5f759cC0F714c94CBc5cc4da",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "PROXY-UPGRADED",
      "ðŸŸ£ CSFeeOracle: Implementation Upgraded",
      "The proxy implementation has been upgraded to 0x4d70efa74ec0ac3a5f759cc0f714c94cbc5cc4da",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0x4D70efa74ec0Ac3a5f759cC0F714c94CBc5cc4da",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;
